File: frontend/src/types/tmdb.ts
/** TMDB API types — aligned with backend `app.schemas.tmdb` + ItemFromTmdbCreate. */

export type TmdbMediaType = "movie" | "tv";
export type TmdbSearchMediaFilter = "all" | "movie" | "tv";
export type TmdbStatusValue = "AVAILABLE" | "NOT_CONFIGURED" | "UNAVAILABLE";
export type TmdbAuthMode = "bearer" | "api_key" | "none";

export interface TmdbStatusResponse {
  status: TmdbStatusValue;
  configured: boolean;
  available: boolean;
  auth_mode: TmdbAuthMode;
  language: string;
  region: string;
  checked_at: string;
}

export interface TmdbGenre {
  id: number;
  name: string;
}

export interface TmdbCastMember {
  tmdb_person_id: number;
  name: string;
  character: string | null;
  profile_path: string | null;
  profile_url: string | null;
  order: number | null;
}

export interface TmdbCrewMember {
  tmdb_person_id: number;
  name: string;
  job: string | null;
  profile_path: string | null;
  profile_url: string | null;
}

export interface TmdbExternalIds {
  imdb_id: string | null;
  wikidata_id: string | null;
  facebook_id: string | null;
  instagram_id: string | null;
  twitter_id: string | null;
}

export interface TmdbSearchResultItem {
  tmdb_id: number;
  media_type: TmdbMediaType;
  title: string;
  original_title: string | null;
  overview: string | null;
  original_language: string | null;
  release_date: string | null;
  release_year: number | null;
  genre_ids: number[];
  poster_path: string | null;
  poster_url: string | null;
  backdrop_path: string | null;
  backdrop_url: string | null;
  adult: boolean | null;
  popularity: number | null;
  vote_average: number | null;
  vote_count: number | null;
  registered: boolean;
  registered_item_id: string | null;
}

export interface TmdbSearchResponse {
  query: string;
  media_type: TmdbSearchMediaFilter;
  page: number;
  upstream_total_pages: number;
  upstream_total_results: number;
  returned_count: number;
  results: TmdbSearchResultItem[];
}

export interface TmdbDetailResponse {
  tmdb_id: number;
  media_type: TmdbMediaType;
  title: string;
  original_title: string | null;
  overview: string | null;
  tagline: string | null;
  original_language: string | null;
  adult: boolean | null;
  status: string | null;
  release_date: string | null;
  release_year: number | null;
  runtime_minutes: number | null;
  episode_runtime_minutes: number[] | null;
  first_air_date: string | null;
  last_air_date: string | null;
  number_of_seasons: number | null;
  number_of_episodes: number | null;
  genres: TmdbGenre[];
  poster_path: string | null;
  poster_url: string | null;
  backdrop_path: string | null;
  backdrop_url: string | null;
  popularity: number | null;
  vote_average: number | null;
  vote_count: number | null;
  cast: TmdbCastMember[];
  directors: TmdbCrewMember[];
  creators: TmdbCrewMember[];
  external_ids: TmdbExternalIds;
  registered: boolean;
  registered_item_id: string | null;
}

export interface ItemFromTmdbCreatePayload {
  media_type: TmdbMediaType;
  tmdb_id: number;
  category_id: string;
  collection_id?: string | null;
  status?: "PLANNED" | "COMPLETED";
  rating?: number;
  progress_note?: string | null;
  memo?: string | null;
  title?: string | null;
}

export interface TmdbAlreadyExistsDetail {
  code: "TMDB_ITEM_ALREADY_EXISTS";
  existing_item_id: string;
}

File: backend/app/schemas/tmdb.py
"""TMDB API response schemas."""

from __future__ import annotations

from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

TmdbMediaType = Literal["movie", "tv"]
TmdbSearchMediaFilter = Literal["all", "movie", "tv"]
TmdbStatusValue = Literal["AVAILABLE", "NOT_CONFIGURED", "UNAVAILABLE"]
TmdbAuthMode = Literal["bearer", "api_key", "none"]


class TmdbStatusResponse(BaseModel):
    status: TmdbStatusValue
    configured: bool
    available: bool
    auth_mode: TmdbAuthMode
    language: str
    region: str
    checked_at: str


class TmdbGenre(BaseModel):
    id: int
    name: str


class TmdbCastMember(BaseModel):
    tmdb_person_id: int
    name: str
    character: str | None = None
    profile_path: str | None = None
    profile_url: str | None = None
    order: int | None = None


class TmdbCrewMember(BaseModel):
    tmdb_person_id: int
    name: str
    job: str | None = None
    profile_path: str | None = None
    profile_url: str | None = None


class TmdbExternalIds(BaseModel):
    imdb_id: str | None = None
    wikidata_id: str | None = None
    facebook_id: str | None = None
    instagram_id: str | None = None
    twitter_id: str | None = None


class TmdbSearchResultItem(BaseModel):
    tmdb_id: int
    media_type: TmdbMediaType
    title: str
    original_title: str | None = None
    overview: str | None = None
    original_language: str | None = None
    release_date: date | None = None
    release_year: int | None = None
    genre_ids: list[int] = Field(default_factory=list)
    poster_path: str | None = None
    poster_url: str | None = None
    backdrop_path: str | None = None
    backdrop_url: str | None = None
    adult: bool | None = None
    popularity: float | None = None
    vote_average: float | None = None
    vote_count: int | None = None
    registered: bool = False
    registered_item_id: UUID | None = None


class TmdbSearchResponse(BaseModel):
    query: str
    media_type: TmdbSearchMediaFilter
    page: int
    upstream_total_pages: int
    upstream_total_results: int
    returned_count: int
    results: list[TmdbSearchResultItem]


class TmdbDetailResponse(BaseModel):
    tmdb_id: int
    media_type: TmdbMediaType
    title: str
    original_title: str | None = None
    overview: str | None = None
    tagline: str | None = None
    original_language: str | None = None
    adult: bool | None = None
    status: str | None = None
    release_date: date | None = None
    release_year: int | None = None
    runtime_minutes: int | None = None
    episode_runtime_minutes: list[int] | None = None
    first_air_date: date | None = None
    last_air_date: date | None = None
    number_of_seasons: int | None = None
    number_of_episodes: int | None = None
    genres: list[TmdbGenre] = Field(default_factory=list)
    poster_path: str | None = None
    poster_url: str | None = None
    backdrop_path: str | None = None
    backdrop_url: str | None = None
    popularity: float | None = None
    vote_average: float | None = None
    vote_count: int | None = None
    cast: list[TmdbCastMember] = Field(default_factory=list)
    directors: list[TmdbCrewMember] = Field(default_factory=list)
    creators: list[TmdbCrewMember] = Field(default_factory=list)
    external_ids: TmdbExternalIds = Field(default_factory=TmdbExternalIds)
    registered: bool = False
    registered_item_id: UUID | None = None
